worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,247 @@
+	Tracef(message string, args ...interface{})
+	IsTraceEnabled() bool
+}
+
+// Stream represents a way to get change events.
+type Stream interface {
+	// Changes returns a channel for a given namespace (database).
+	Changes() <-chan changestream.ChangeEvent
+}
+
+type subscription struct {
+	id      int
+	changes chan changestream.ChangeEvent
+	topics  map[string]struct{}
+
+	unsubscriber func()
+}
+
+// Unsubscribe removes the subscription from the event queue.
+func (s *subscription) Unsubscribe() {
+	s.unsubscriber()
+}
+
+// Changes returns a channel that will receive events from the event queue
+// for the subscription.
+func (s *subscription) Changes() <-chan changestream.ChangeEvent {
+	return s.changes
+}
+
+type eventFilter struct {
+	subscriptionID int
+	changeMask     changestream.ChangeType
+	filter         func(changestream.ChangeEvent) bool
+}
+
+// EventQueue is a queue of events that can be subscribed to.
+type EventQueue struct {
+	tomb   tomb.Tomb
+	stream Stream
+	logger Logger
+
+	mutex              sync.Mutex
+	subscriptions      map[int]*subscription
+	subscriptionsByNS  map[string][]*eventFilter
+	subscriptionsCount int
+
+	actions chan func()
+}
+
+// New creates a new EventQueue that will use the Stream for events.
+func New(stream Stream, logger Logger) *EventQueue {
+	queue := &EventQueue{
+		stream:             stream,
+		logger:             logger,
+		subscriptions:      make(map[int]*subscription),
+		subscriptionsByNS:  make(map[string][]*eventFilter),
+		subscriptionsCount: 0,
+		actions:            make(chan func()),
+	}
+
+	queue.tomb.Go(queue.loop)
+
+	return queue
+}
+
+// Subscribe creates a new subscription to the event queue.
+func (s *EventQueue) Subscribe(opts ...changestream.SubscriptionOption) (changestream.Subscription, error) {
+	if len(opts) == 0 {
+		return nil, errors.Errorf("no subscription options specified")
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	// Create a new subscription and assign a unique ID to it.
+	s.subscriptionsCount++
+	sub := &subscription{
+		id:           s.subscriptionsCount,
+		changes:      make(chan changestream.ChangeEvent),
+		topics:       make(map[string]struct{}),
+		unsubscriber: func() { s.unsubscribe(s.subscriptionsCount) },
+	}
+	s.subscriptions[sub.id] = sub
+
+	// Register filters to route changes matching the subscription criteria to the newly crated subscription.
+	for _, opt := range opts {
+		namespace := opt.Namespace()
+		s.subscriptionsByNS[namespace] = append(s.subscriptionsByNS[namespace], &eventFilter{
+			subscriptionID: sub.id,
+			changeMask:     opt.ChangeMask(),
+			filter:         opt.Filter(),
+		})
+		sub.topics[namespace] = struct{}{}
+	}
+
+	return sub, nil
+}
+
+// Kill stops the event queue.
+func (s *EventQueue) Kill() {
+	s.tomb.Kill(nil)
+}
+
+// Wait waits for the event queue to stop.
+func (s *EventQueue) Wait() error {
+	return s.tomb.Wait()
+}
+
+func (s *EventQueue) unsubscribe(subscriptionID int) {
```

Comment:
You could push this logic into the event loop, then you don't race with dispatch, and you don't need to reason about mutex holding as much.

Reply:
I did try this originally by serializing the actions, but there was a race somewhere.